src/stdio/src/helpers.rs

```diff
@@ -0,0 +1,159 @@
+use super::{internal, BUFSIZ, FILE, UNGET};
+use compiler_builtins::mem::memset;
+use stdlib::calloc;
+use core::{mem, ptr};
+use core::sync::atomic::AtomicBool;
+use platform::types::*;
+use super::constants::*;
+use fcntl::*;
+use platform;
+use errno;
+
+/// Parse mode flags as a string and output a mode flags integer
+pub unsafe fn parse_mode_flags(mode_str: *const c_char) -> i32 {
+    use string::strchr;
+    let mut flags = if !strchr(mode_str, b'+' as i32).is_null() {
+        O_RDWR
+    } else if (*mode_str) == b'r' as i8 {
+        O_RDONLY
+    } else {
+        O_WRONLY
+    };
+    if !strchr(mode_str, b'x' as i32).is_null() {
+        flags |= O_EXCL;
+    }
+    if !strchr(mode_str, b'e' as i32).is_null() {
+        flags |= O_CLOEXEC;
+    }
+    if (*mode_str) != b'r' as i8 {
+        flags |= O_CREAT;
+    }
+    if (*mode_str) == b'w' as i8 {
+        flags |= O_TRUNC;
+    }
+    if (*mode_str) != b'a' as i8 {
+        flags |= O_APPEND;
+    }
+
+    flags
+}
+
+/// Open a file with the file descriptor `fd` in the mode `mode`
+pub unsafe fn _fdopen(fd: c_int, mode: *const c_char) -> *mut FILE {
+    use string::strchr;
+    if *mode != b'r' as i8 && *mode != b'w' as i8 && *mode != b'a' as i8 {
+        platform::errno = errno::EINVAL;
+        return ptr::null_mut();
+    }
+
+    let mut flags = 0;
+    if strchr(mode, b'+' as i32).is_null() {
+        flags |= if *mode == b'r' as i8 { F_NOWR } else { F_NORD };
+    }
+
+    if !strchr(mode, b'e' as i32).is_null() {
+        sys_fcntl(fd, F_SETFD, FD_CLOEXEC);
+    }
+
+    if *mode == 'a' as i8 {
+        let f = sys_fcntl(fd, F_GETFL, 0);
+        if (f & O_APPEND) == 0 {
+            sys_fcntl(fd, F_SETFL, f | O_APPEND);
+        }
+        flags |= F_APP;
+    }
+
+    let file = calloc(mem::size_of::<FILE>() + BUFSIZ + UNGET, 1) as *mut FILE;
+    // Allocate the file
+    (*file) = FILE {
+        flags: flags,
+        rpos: ptr::null_mut(),
+        rend: ptr::null_mut(),
+        wend: ptr::null_mut(),
+        wpos: ptr::null_mut(),
+        wbase: ptr::null_mut(),
+        fd: fd,
+        buf: (file as *mut u8).add(mem::size_of::<FILE>() + UNGET),
+        buf_size: BUFSIZ,
+        buf_char: -1,
+        unget: UNGET,
+        lock: AtomicBool::new(false),
+        write: Some(&internal::stdio_write),
+        read: Some(&internal::stdio_read),
+        seek: Some(&internal::stdio_seek),
+    };
+    file
+}
+
+/// Write buffer `buf` of length `l` into `stream`
+pub unsafe fn fwritex(buf: *const u8, l: size_t, stream: *mut FILE) -> size_t {
+    use compiler_builtins::mem::memcpy;
+    let mut buf = buf;
+    let mut l = l;
+    if let Some(stream_write) = (*stream).write {
```

Comment:
We could, but I don't see how it would help. If the `FILE *` being passed into the functions is not valid, then that should be the programmers fault, and we should not be catching that. If you pass a `NULL` to any of the stdio functions in glibc or musl, you will get a null pointer dereference error, as they don't do any checking.

Reply:
:+1: